File: Top-Interview-Questions-Leetcode/Easy-Collections/trees/convert-sorted-array-to-BST.cpp
#include <iostream>
#include <vector>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    
    TreeNode* sortedArrayToBST(std::vector<int>& nums, int l, int h) {
        if (l > h) return nullptr;
        
        if (l == h) {
            TreeNode* node = new TreeNode(nums[l]);
            return node;
        }
        
        int mid = l + (h - l) / 2;
        
        TreeNode* left = sortedArrayToBST(nums, l, mid - 1);
        TreeNode* node = new TreeNode(nums[mid]);
        TreeNode* right = sortedArrayToBST(nums, mid + 1, h);
        
        node->left = left;
        node->right = right;
        
        return node;
    }
    
public:
    TreeNode* sortedArrayToBST(std::vector<int>& nums) {
        return sortedArrayToBST(nums, 0, nums.size() - 1);
    }
};

File: Top-Interview-Questions-Leetcode/Easy-Collections/trees/convert-sorted-array-to-BST.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __sortedArrayToBST(self, nums: List[int], l: int, h: int) -> Optional[TreeNode]:
        if l > h:
            return None
        if l == h:
            node = TreeNode(nums[l])
            return node
        
        m = l + (h - l) // 2
        
        left = self.__sortedArrayToBST(nums, l, m - 1)
        node = TreeNode(nums[m])
        right = self.__sortedArrayToBST(nums, m + 1, h)
        
        node.left = left
        node.right = right
        
        return node
    
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        return self.__sortedArrayToBST(nums, 0, len(nums) - 1)